flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceSplitEnumerator.java

```diff
@@ -361,7 +394,7 @@ public Map<Integer, ReaderInfo> registeredReaders() {
             }
             Integer lastIndex = null;
             for (Integer sourceIndex : readerSourceIndex.values()) {
-                if (lastIndex != null && lastIndex != sourceIndex) {
```

Comment:
We may want to merge this fix as a separate change since it had been reported in the past and is unrelated to the race condition.